lib/fp.js

```diff
@@ -1,51 +1,37 @@
 'use strict';
 
-let asyncChainMethods = null;
-// Convert synchronous function to asynchronous
-// Transform function with args arguments and callback
-// to function with args as separate values and callback
-//   fn - <Function>, callback-last / err-first
-//
-// Returns: <Function>
-const toAsync = fn => (...argsCb) => {
-  const len = argsCb.length - 1;
-  const callback = argsCb[len];
-  const args = argsCb.slice(0, len);
-  return fn(args, callback);
-};
-
-// Wrap function adding async chain methods
-//   fn - <Function>, asynchronous
-//   args - <Array>, its arguments
-const asAsync = (fn, ...args) => {
+// Add functional methods to `fn` and partially apply arguments to it.
+// Functional methods are added by replacing prototype of `fn`.
+//   fn - <Function> function which functional methods will be added to
+//   args - <any[]> arguments which will be partially applied to fn
+const functorify = (fn, ...args) => {
   const wrapped = fn.bind(null, ...args);
-  for (const name in asyncChainMethods) {
-    const method = asyncChainMethods[name];
-    wrapped[name] = (...args) => asAsync(method(wrapped, ...args));
-  }
+  Object.setPrototypeOf(wrapped, Functor.prototype);
   return wrapped;
 };
 
 // Applicative f => a -> f a
 //   args - <Array>
-const of = (...args) => asAsync(callback => callback(null, ...args));
+const of = (...args) => functorify(callback => callback(null, ...args));
 
 // Monoid m => a -> a -> a
```

Comment:
`m a -> m a -> m a`?

Reply:
Thanks for point out. I think it should be `Monoid m => m -> m -> m`